docs/spec_draft.md

```diff
@@ -1267,6 +1268,70 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Applies a `body` function to `inputs` and `init_values` along the `dimensions`
+and produces a `result` tensor.
+
+The `body` function has to be associative, and it produces an
+implementation-defined behavior otherwise. Similarly, if `init_values` do not
+form identity under the `body` function, it produces an implementation-defined
+behavior.
```

Comment:
And here's the formalization (I'm surprised at how compact it turned out):

```
More formally, `results[k][result_index] = reduce(input_slices)[k]` for all `k`
$\in$ [0, N), where:
  * `input_slices[k]` = `inputs[k][ri0, ..., :, ..., riR-1]` for all `k` 
    $\in$ [0, N), where `ri` are individual elements in `result_index` and `:` 
    are inserted at `dimensions`.
  * `reduce(input_slices)` = `eval(schedule)` for some binary tree `schedule` 
    where:
      * `eval(node)` = `body(eval(node.left), eval(node.right))`.
      * `eval(leaf)` = `leaf.value`.
  * `schedule` is an implementation-defined full binary tree whose in-order
    traversal consists of:
      * `input_slices[:][index]` values, for all `index` in the index space
        of `input_slices`, in the ascending lexicographic order of `index`.
      * Interspersed with an implementation-defined amount of `init_values`
        at implementation-defined positions.
```

Reply:
> `schedule` is an implementation-defined full binary tree whose in-order
    traversal consists of:

Though the actual traversal order is implementation-defined (and whether interpreter should compute in parallel), shouldn't we say reverse in-order traversal since we want to compute the bottom-most nodes first if we want it to be computed in parallel:
reduce sum: `[1, 2, 3, 4]`
`f(f(f(f(i, 1), 2), 3), 4)`
```
                   f(f(f(f(i, 1), 2), 3), 4)
                      /                   \
      f(f(f(i, 1), 2), 3)                  4
         /                \
  f(f(i, 1), 2)            3
     /    \
f(i, 1)    2
  /  \ 
 i    1 
```
vs
`f(f(f(i, 1), f(i, 2)), f(f(i, 3), f(i, 4)))`
```
    f(f(f(i, 1), f(i, 2)), f(f(i, 3), f(i, 4)))
           /                            \
   f(f(i, 1), f(i, 2))        f(f(i, 3), f(i, 4))
     /        \                       /        \
 f(i, 1)   f(i, 2)            f(i, 3)        f(i, 4)
  /  \        /  \                /  \         /  \
 i    1      i    2              i    3       i    4
```

Reply:
This is what I understood from the formalism:
1. `input_slices = ...`. The order of elements in `input_slices`, which depends on how `result_index` are visited, is not mentioned. 
2. Given a particular order of elements in `input_slices`, there could be multiple binary trees following the same order via inorder traversal. It's implementation defined which one of those are picked. In the example above, both the trees are fine to be created by two different implementations as they have the same in-order traversals.  

4. Given a particular tree, the order in which the nodes (including non-leaf nodes) are evaluated is related more to how someone wants to implement it: which does not matter in the formulation. Maybe a non-leaf node can be evaluated partially before moving on to the some other and can be resumed later once its dependencies are met or evaluation  follows a recursive top down approach where the leafs will be evaluated first (as you pointed out). 

I think the reason for (1) is that the order does not matter. Now (2) is based on that order. My question is why we need the full binary tree schedule to follow the order in (1) via in-order traversal. Why not any just full binary tree with the elements of `input_slice`?

Reply:
"It's implementation defined which one of those are picked. In the example above, both the trees are fine to be created by two different implementations as they have the same in-order traversals". Yes, that's what I had in mind in the proposed formulation. The in-order traversal called out above is only used to constrain the order in which the leaves are laid out in these potential trees and is unrelated the traversal used to evaluate a particular tree.

"My question is why we need the full binary tree schedule to follow the order in (1) via in-order traversal. Why not any just full binary tree with the elements of input_slice?". I think that's because we don't know if the reduction operation is commutative or not. So it's fine to do `(a op b) op (c op d)`, `((a op b) op c) op d)` etc. But it's not fine to do e.g. `(b op a) op (c op d)`.